File: pixijs/render/room-bonus-vfx.js
// ============================================================
// ROOM BONUS VFX — ambient floating emoji particles in bonus rooms
//
// initRoomBonusVfx(layer)   — create container on given layer
// updateRoomBonusVfx(state) — spawn + animate emoji per frame
// clearRoomBonusVfx()       — destroy all, clear pool
//
// Globals: CONFIG, ROOM_BONUS_TYPES (from config.js)
// ============================================================

import { Container, Text } from 'pixi.js';
import { CELL_PX } from '../world/constants.js';

const POOL_SIZE = 30;
const SPAWN_INTERVAL = 0.4;
const LIFE_MIN = 2.0;
const LIFE_MAX = 3.0;
const FONT_SIZE = 12;

let _container = null;
const _pool = [];   // idle Text objects
const _active = []; // { spr, vx, vy, life, maxLife, baseX, wobblePhase, wobbleAmp }

// ── Public API ────────────────────────────────────────────────

export function initRoomBonusVfx(layer) {
  _container = new Container();
  layer.addChild(_container);
  for (let i = 0; i < POOL_SIZE; i++) _pool.push(_makeText());
}

export function clearRoomBonusVfx() {
  if (!_container) return;
  for (const a of _active) {
    a.spr.visible = false;
    _pool.push(a.spr);
  }
  _active.length = 0;
  _container.removeChildren();
}

let _timer = 0;

export function updateRoomBonusVfx(state, dt) {
  if (!_container) return;
  if (state.phase !== 'play') {
    // hide all active when not in play
    for (const a of _active) {
      a.spr.visible = false;
      _pool.push(a.spr);
    }
    _active.length = 0;
    _container.removeChildren();
    return;
  }

  // Spawn
  _timer += dt;
  if (_timer >= SPAWN_INTERVAL) {
    _timer = 0;
    _spawn(state);
  }

  // Update active
  for (let i = _active.length - 1; i >= 0; i--) {
    const a = _active[i];
    a.life -= dt;
    if (a.life <= 0) {
      a.spr.visible = false;
      _pool.push(a.spr);
      _active.splice(i, 1);
      _container.removeChild(a.spr);
      continue;
    }
    a.baseX += a.vx * dt;
    a.spr.y += a.vy * dt;
    a.wobblePhase += dt * 2;
    a.spr.x = a.baseX + Math.sin(a.wobblePhase) * a.wobbleAmp;

    // kill if drifted outside the cell bounds
    const cellMinX = a.cellX * CELL_PX;
    const cellMaxX = (a.cellX + 1) * CELL_PX;
    const cellMinY = a.cellY * CELL_PX;
    const cellMaxY = (a.cellY + 1) * CELL_PX;
    if (a.spr.x < cellMinX || a.spr.x > cellMaxX ||
        a.spr.y < cellMinY || a.spr.y > cellMaxY) {
      a.life = 0;
    }

    // alpha fade: fade in first 20%, fade out last 30%
    const t = 1 - a.life / a.maxLife;
    let alpha;
    if (t < 0.2) alpha = t / 0.2;
    else if (t > 0.7) alpha = (1 - t) / 0.3;
    else alpha = 1;
    a.spr.alpha = alpha;

    // gentle scale pulse
    const scale = 1 + Math.sin(a.wobblePhase * 1.5) * 0.1;
    a.spr.scale.set(scale);
  }
}

// ── Internal ──────────────────────────────────────────────────

function _spawn(state) {
  const bonuses = state.roomBonuses;
  if (!bonuses || bonuses.length === 0) return;
  const rooms = state.rooms;
  if (!rooms) return;
  const everRevealed = state.everRevealedCells;
  if (!everRevealed) return;

  for (const rb of bonuses) {
    const room = rooms[rb.roomIdx];
    if (!room || !room.cells || room.cells.length === 0) continue;

    // pick a random cell that's been revealed
    const revealedCells = room.cells.filter(c => everRevealed.has(c.k));
    if (revealedCells.length === 0) continue;
    const cell = revealedCells[Math.floor(Math.random() * revealedCells.length)];

    const bonusDef = (typeof ROOM_BONUS_TYPES !== 'undefined')
      ? ROOM_BONUS_TYPES.find(bt => bt.id === rb.bonusType)
      : null;
    if (!bonusDef || !bonusDef.icon) continue;

    const spr = _pool.pop() ?? _makeText();
    spr.text = bonusDef.icon;
    spr.style.fontSize = FONT_SIZE;
    spr.anchor.set(0.5);

    // random position within the cell (with some margin)
    const margin = CELL_PX * 0.2;
    const px = (cell.x + 0.5) * CELL_PX + (Math.random() - 0.5) * (CELL_PX - margin * 2);
    const py = (cell.y + 0.5) * CELL_PX + (Math.random() - 0.5) * (CELL_PX - margin * 2);
    spr.x = px;
    spr.y = py;
    spr.alpha = 0;
    spr.visible = true;
    _container.addChild(spr);

    const life = LIFE_MIN + Math.random() * (LIFE_MAX - LIFE_MIN);
    _active.push({
      spr,
      vx: (Math.random() - 0.5) * 10,
      vy: -5 - Math.random() * 8,
      life,
      maxLife: life,
      baseX: px,
      cellX: cell.x,
      cellY: cell.y,
      wobblePhase: Math.random() * Math.PI * 2,
      wobbleAmp: 3 + Math.random() * 4,
    });
  }
}

function _makeText() {
  return new Text({
    text: '',
    style: { fontSize: FONT_SIZE, fontFamily: 'sans-serif' },
  });
}

File: Scripts/Presentation/RoomBonusVfx.cs
using System.Collections.Generic;
using Godot;
using SpaceOrLife.Core;
using SpaceOrLife.Core.Catalogs;
using SpaceOrLife.Domain;
using SpaceOrLife.World;

namespace SpaceOrLife.Presentation;

public partial class RoomBonusVfx : Node2D
{
    private const int PoolSize = 30;
    private const float SpawnInterval = 0.4f;
    private const float LifeMin = 2.0f;
    private const float LifeMax = 3.0f;
    private const int FontSize = 12;

    private struct ActiveVfx
    {
        public Label Label;
        public float Vx;
        public float Vy;
        public float Life;
        public float MaxLife;
        public float BaseX;
        public int CellX;
        public int CellY;
        public float WobblePhase;
        public float WobbleAmp;
    }

    private readonly List<Label> _pool = new();
    private readonly List<ActiveVfx> _active = new();
    private float _timer;
    private LevelState? _state;
    private RoomBonusCatalog? _catalog;
    private bool _active_phase;

    public override void _Ready()
    {
        for (int i = 0; i < PoolSize; i++)
        {
            var label = new Label
            {
                Visible = false,
                ZIndex = 50,
            };
            label.AddThemeFontSizeOverride("font_size", FontSize);
            label.AddThemeColorOverride("font_color", Colors.White);
            AddChild(label);
            _pool.Add(label);
        }
    }

    public void SetState(LevelState state, RoomBonusCatalog catalog)
    {
        _state = state;
        _catalog = catalog;
    }

    public void SetActive(bool active)
    {
        _active_phase = active;
        if (!active && _active.Count > 0)
            Clear();
    }

    public void Clear()
    {
        for (int i = _active.Count - 1; i >= 0; i--)
        {
            var a = _active[i];
            a.Label.Visible = false;
            _pool.Add(a.Label);
        }
        _active.Clear();
    }

    public override void _Process(double delta)
    {
        var dt = (float)delta;

        if (_state == null || _catalog == null || !_active_phase)
        {
            if (_active.Count > 0)
                Clear();
            return;
        }

        _timer += dt;
        if (_timer >= SpawnInterval)
        {
            _timer = 0f;
            Spawn();
        }

        var cellPx = GameConstants.CellPx;

        for (int i = _active.Count - 1; i >= 0; i--)
        {
            var a = _active[i];
            a.Life -= dt;

            if (a.Life <= 0f)
            {
                a.Label.Visible = false;
                _pool.Add(a.Label);
                _active.RemoveAt(i);
                continue;
            }

            a.BaseX += a.Vx * dt;
            a.Label.Position = new Vector2(a.BaseX, a.Label.Position.Y + a.Vy * dt);
            a.WobblePhase += dt * 2f;
            a.Label.Position = new Vector2(a.BaseX + Mathf.Sin(a.WobblePhase) * a.WobbleAmp, a.Label.Position.Y);

            var cellMinX = a.CellX * cellPx;
            var cellMaxX = (a.CellX + 1) * cellPx;
            var cellMinY = a.CellY * cellPx;
            var cellMaxY = (a.CellY + 1) * cellPx;
            if (a.Label.Position.X < cellMinX || a.Label.Position.X > cellMaxX ||
                a.Label.Position.Y < cellMinY || a.Label.Position.Y > cellMaxY)
            {
                a.Life = 0f;
            }

            var t = 1f - a.Life / a.MaxLife;
            float alpha;
            if (t < 0.2f)
                alpha = t / 0.2f;
            else if (t > 0.7f)
                alpha = (1f - t) / 0.3f;
            else
                alpha = 1f;
            a.Label.Modulate = new Color(1f, 1f, 1f, alpha);

            var scale = 1f + Mathf.Sin(a.WobblePhase * 1.5f) * 0.1f;
            a.Label.Scale = new Vector2(scale, scale);

            _active[i] = a;
        }
    }

    private void Spawn()
    {
        if (_state == null || _catalog == null) return;
        var bonuses = _state.RoomBonuses;
        if (bonuses.Count == 0) return;
        var rooms = _state.Rooms;
        if (rooms.Count == 0) return;
        var everRevealed = _state.EverRevealedCells;

        var cellPx = GameConstants.CellPx;

        foreach (var rb in bonuses)
        {
            if (rb.RoomIdx < 0 || rb.RoomIdx >= rooms.Count) continue;
            var room = rooms[rb.RoomIdx];
            if (room.Cells.Count == 0) continue;

            var revealedCells = new List<CellCoord>();
            foreach (var c in room.Cells)
            {
                if (everRevealed.Contains(c))
                    revealedCells.Add(c);
            }
            if (revealedCells.Count == 0) continue;

            var def = _catalog.GetById(rb.BonusType);
            if (def == null || string.IsNullOrEmpty(def.Icon)) continue;

            var cell = revealedCells[GD.RandRange(0, revealedCells.Count - 1)];
            var margin = cellPx * 0.2f;
            var px = (cell.X + 0.5f) * cellPx + (float)GD.RandRange(-(cellPx - margin * 2) * 0.5, (cellPx - margin * 2) * 0.5);
            var py = (cell.Y + 0.5f) * cellPx + (float)GD.RandRange(-(cellPx - margin * 2) * 0.5, (cellPx - margin * 2) * 0.5);

            if (_pool.Count == 0) return;
            var label = _pool[_pool.Count - 1];
            _pool.RemoveAt(_pool.Count - 1);

            label.Text = def.Icon;
            label.Position = new Vector2(px, py);
            label.Modulate = new Color(1f, 1f, 1f, 0f);
            label.Scale = Vector2.One;
            label.Visible = true;

            var life = LifeMin + (float)GD.RandRange(0, LifeMax - LifeMin);
            _active.Add(new ActiveVfx
            {
                Label = label,
                Vx = (float)GD.RandRange(-5, 5),
                Vy = -5f - (float)GD.RandRange(0, 8),
                Life = life,
                MaxLife = life,
                BaseX = px,
                CellX = cell.X,
                CellY = cell.Y,
                WobblePhase = (float)GD.RandRange(0, Mathf.Tau),
                WobbleAmp = 3f + (float)GD.RandRange(0, 4),
            });
        }
    }
}
